libpromises/evalfunction.c

```diff
@@ -8451,6 +8452,231 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
 
 /*********************************************************************/
 
+struct IsReadableThreadData
+{
+    pthread_t thread;
+    pthread_cond_t cond;
+    pthread_mutex_t mutex;
+    const char *path;
+    FnCallResult result;
+};
+
+static void *IsReadableThreadRoutine(void *data)
+{
+    assert(data != NULL);
+
+    struct IsReadableThreadData *const thread_data = data;
+
+    int ret = pthread_mutex_lock(&thread_data->mutex);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to lock mutex: %s",
+            GetErrorStrFromCode(ret));
+        return NULL;
+    }
+
+    thread_data->result = FnFailure();
+
+    ret = pthread_mutex_unlock(&thread_data->mutex);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to unlock mutex: %s",
+            GetErrorStrFromCode(ret));
+        return NULL;
+    }
+
+    bool result = false;
+    char buf[1];
+    const int fd = open(thread_data->path, O_RDONLY);
+    if (fd < 0) {
+        Log(LOG_LEVEL_DEBUG, "Failed to open file '%s': %s",
+            thread_data->path, GetErrorStr());
+    }
+    else if (read(fd, buf, sizeof(buf)) < 0)
+    {
+        Log(LOG_LEVEL_DEBUG, "Failed to read from file '%s': %s",
+            thread_data->path, GetErrorStr());
+        close(fd);
+    }
+    else
+    {
+        close(fd);
+        result = true;
+    }
+
+    ret = pthread_mutex_lock(&thread_data->mutex);
```

Comment:
Why do we need this second lock-unlock?

Reply:
We don't need it. I was initially thinking there could be a race condition where the main thread reads the result while it is being assigned. But the main thread only reads the variable upon receiving a signal on the condition.